File: 8Puzzle/newGame.py
from __future__ import print_function
import math

start = [[2,8,3], 
         [1,6,4], 
         [0,7,5]]

curup = [[2,8,3], 
         [1,6,4], 
         [0,7,5]]

curdown = [[2,8,3], 
         [1,6,4], 
         [0,7,5]]

curleft = [[2, 8, 3],
         [1, 6, 4],
         [0, 7, 5]]

curright = [[2, 8, 3],
         [1, 6, 4],
         [0, 7, 5]]

goal =  [[1,2,3], 
         [8,0,4], 
         [7,6,5]]

def display(state):
  for i in state:
    for j in i:
      print(j, end=" ")
    print()

def getIndex(state, elemant):
  for i in range(3):
    for j in range(3):
      if state[i][j] is elemant:
        yield i
        yield j
        break

def swap(state, i, j, direction):
  temp = state[i][j]
  if direction is 'd' and i is not 0:
    state[i][j] = state[i-1][j]
    state[i-1][j] = temp

  elif direction is 'u' and i is not 2:
    state[i][j] = state[i+1][j]
    state[i+1][j] = temp
  
  elif direction is 'r' and j is not 0:
    state[i][j] = state[i][j-1]
    state[i][j-1] = temp
    
  elif direction is 'l' and j is not 2:
    state[i][j] = state[i][j+1]
    state[i][j+1] = temp    
  return state

def move(state, direction):
  i, j = getIndex(state, 0)
  state = swap(state, i, j, direction)
  #display(state)
  #print(getFScore(state,0))

def getHScore(state):
  HScore = 0
  for i in range(3):
    for j in range(3):
      elemant = state[i][j]
      if elemant is not 0:
       a, b = getIndex(goal,elemant)
       HScore = HScore + abs(i-a) + abs(j-b)
  return HScore

def getGScore(state, prevG):
  GScore = 0
  for i in range(3):
    for j in range(3):
      if goal[i][j] is not state[i][j]:
        GScore += 1
  return prevG + GScore

def getFScore(state, prevF):
  return getHScore(state) + getGScore(state, prevF)

def minn(listCost, up, down, left, right):
  min = 100000
  for i in listCost:
    if i < min:
      min = i

  if min is up:
    return 'u'
  if min is down:
    return 'd'
  if min is left:
    return 'l'
  if min is right:
    return 'r'

close = []
close.append(start)
i=-1
while start is not goal:
  i+=1
  up = down = left = right = 100000
  move(curdown,'d')
  move(curup,'u')
  move(curleft,'l')
  move(curright,'r')

  if curdown not in close:
    down = getFScore(curdown,0)
  if curup not in close:
    up = getFScore(curup,0)
  if curleft not in close:
    left = getFScore(curleft,0)
  if curright not in close:
    right = getFScore(curright, 0)
  listCost = [up, down, left, right]
  s = minn(listCost, up, down, left, right)
  #print(s, end=" ")
  move(start, s)
  close.append([])
  close[i].append(start)
  print()
  close.append(start)

File: 8Puzzle/newGame.java
import java.util.*;
class newGame
{
  public static int[][] start = new int[][] {
    {1,2,3}, 
    {4,0,6}, 
    {7,8,5}};
  
  public static int[][] goal = new int[][] {
    {1,2,3}, 
    {4,5,6}, 
    {7,0,8}};

  public static int[][][] closeSet = new int[100][3][3];
  public static Scanner sc = new Scanner(System.in);

  static void display(int[][][] arr) {
    System.out.println();
    for (int k = 0; k < 100; k++) {
      for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
          System.out.print(" " + arr[k][i][j]);
        }
        System.out.println();
      }
    }
  }

  static void display(int[][] arr) {
    System.out.println();
    for(int i=0; i<3; i++)
    {
      for(int j=0; j<3; j++)
      {
  
          System.out.print(" "+arr[i][j]);
      }
      System.out.println();
    }
  }

  static int[] getIndex(int[][] arrGoal, int find) {
    int[] indexes = new int[2];

    for (int i = 0; i < 3; i++) 
    {
      for (int j = 0; j < 3; j++) 
      {
        if(arrGoal[i][j] == find)
        {
          indexes[0] = i;
          indexes[1] = j;
          return indexes;
        }
      }
    }
    return indexes;
  }

  static int getHScore(int[][] arr) {
    int hScore = 0;

    for (int i = 0; i < 3; i++) 
    {
      for (int j = 0; j < 3; j++) 
      {
        if(arr[i][j]!=0)
        {
          int[] index = getIndex(goal,arr[i][j]);
        
          hScore += Math.abs(i- index[0]) + Math.abs(j- index[1]);
        }
      }
    }
    return hScore;
  }

  static int getGScore(int[][] arr) {
    int gScore = 0;

    for(int i=0; i<3; i++)
    {
      for(int j=0; j<3; j++)
      {
        if(arr[i][j] != goal[i][j] && arr[i][j] != 0)
        {
          gScore++;
        }
      }
    }
    return gScore;
  }

  static int finalScore(int[][] arr) {
    return getGScore(arr) + getHScore(arr);
  }

  static int[][] move(int[][] arr, char direction) {
    int[][] newState = arr;
    int[] index = getIndex(arr, 0);
    int temp = 0;

    if(direction == 'u' && index[0] != 2)
    {
      newState[index[0]][index[1]] = newState[index[0]+1][index[1]];
      newState[index[0]+1][index[1]] = 0;
    }
    if(direction == 'd' && index[0] != 0)
    {
      newState[index[0]][index[1]] = newState[index[0]-1][index[1]];
      newState[index[0]-1][index[1]] = 0;
    }
    if (direction == 'l' && index[1] != 2) 
    {
      newState[index[0]][index[1]] = newState[index[0]][index[1] + 1];
      newState[index[0]][index[1] + 1] = 0;
    }
    if(direction == 'r' && index[1] != 0)
    {
      newState[index[0]][index[1]] = newState[index[0]][index[1] - 1];
      newState[index[0]][index[1] - 1] = 0;
    }

    return newState;
  }

  static boolean isVisited(int[][] arr)
  {
    for(int k=0; k<100; k++)
    {
      int c = 0;
      for(int i=0; i<3; i++)
      {
        for(int j=0; j<3; j++)
        {
          if(closeSet[k][i][j] == arr[i][j])
            c++;
        }
      }
      if(c==9)
      {
        return true;
      }  
    }
    return false;
  }

  public static void main(String args[])
  {
    
    int[][] state = start;
    int[][] stateUp = start;
    int[][] stateDown = start;
    int[][] stateLeft = start;
    int[][] stateRight = start;
    
    display(state);
    stateUp = move(stateUp, 'u');
    display(stateUp);
    stateDown = move(stateDown, 'd');
    display(stateDown);
    stateLeft = move(stateLeft, 'l');
    display(stateLeft);
    stateRight = move(stateRight, 'r');
      
    
    
    
    
    display(stateRight);
    
  }
}